transport/peeragent.go

```diff
@@ -0,0 +1,13 @@
+package transport
+
+//go:generate mockgen -destination=transporttest/peeragent.go -package=transporttest go.uber.org/yarpc/transport PeerAgent
+
+/*
+PeerAgent represents an interface for managing Peers across different outbounds.  A PeerList will
+request a Peer for a specific PeerIdentifier and the PeerAgent has the ability to create a new Peer or
+return an existing one.
+*/
+type PeerAgent interface {
+	RetainPeer(PeerIdentifier, PeerSubscriber) (Peer, error) // Get or create a Peer for the PeerSubscriber
```

Comment:
So the idea was that:

1) Peers can be contained by multiple subscribers(PeerLists)
2) Peers have a list of their subscribers
3) When actions happen on a peer (Connect/Available/Unavailable/Pending changed) they notify their subscriptions
4) Agents Retain/Release Subscriptions on Peers
5) If a peer has no subscriptions it is safe to delete (from the Agent)